modules/storytelling/story_tellers/close_to_junction_teller.cc

```diff
@@ -18,22 +18,96 @@
 #include "modules/planning/proto/planning.pb.h"
+#include "modules/prediction/common/prediction_gflags.h"
+#include "modules/prediction/common/prediction_map.h"
 #include "modules/storytelling/frame_manager.h"
 
 namespace apollo {
 namespace storytelling {
+namespace {
 
 using apollo::planning::ADCTrajectory;
+using apollo::prediction::PredictionMap;
+using apollo::common::PathPoint;
+
+bool IsPointInPNCJunction(const PathPoint& point) {
+  const auto junctions = PredictionMap::GetPNCJunctions(
+      Eigen::Vector2d(point.x(), point.y()), FLAGS_junction_search_radius);
+  if (junctions.empty() || junctions.front() == nullptr) {
+    return false;
+  }
+  const auto& junction_info = junctions.front();
+  return junction_info != nullptr &&
+         junction_info->pnc_junction().polygon().point_size() >= 3;
+}
+
+bool IsPointInRegularJunction(const PathPoint& point) {
+  const auto junctions = PredictionMap::GetJunctions(
+      Eigen::Vector2d(point.x(), point.y()), FLAGS_junction_search_radius);
+  if (junctions.empty() || junctions.front() == nullptr) {
+    return false;
+  }
+  const auto& junction_info = junctions.front();
+  return junction_info != nullptr &&
+         junction_info->junction().polygon().point_size() >= 3;
+}
+
+double DistanceToPNCJunction(const ADCTrajectory& adc_trajectory) {
+  const double s_start = adc_trajectory.trajectory_point(0).path_point().s();
+  // Test for PNCJunction.
+  for (const auto& point : adc_trajectory.trajectory_point()) {
+    const auto& path_point = point.path_point();
+    if (path_point.s() > FLAGS_adc_trajectory_search_length) {
+      break;
+    }
+    if (IsPointInPNCJunction(path_point)) {
+      return path_point.s() - s_start;
+    }
+  }
+
+  // Test for regular junction.
+  for (const auto& point : adc_trajectory.trajectory_point()) {
+    const auto& path_point = point.path_point();
+    if (path_point.s() > FLAGS_adc_trajectory_search_length) {
+      break;
+    }
+    if (IsPointInRegularJunction(path_point)) {
+      return path_point.s() - s_start;
+    }
+  }
+
+  return -1;
+}
+
+}  // namespace
 
 void CloseToJunctionTeller::Init() {
   auto* manager = FrameManager::Instance();
   manager->CreateOrGetReader<ADCTrajectory>(FLAGS_planning_trajectory_topic);
 }
 
 void CloseToJunctionTeller::Update(Stories* stories) {
-  // TODO(xiaoxq): Implement.
-  // Log 'Enter CloseToJunction Story'
-  // Log 'Exit CloseToJunction Story'
+  auto* manager = FrameManager::Instance();
+  static auto planning_reader = manager->CreateOrGetReader<ADCTrajectory>(
+      FLAGS_planning_trajectory_topic);
+  const auto trajectory = planning_reader->GetLatestObserved();
+  if (trajectory == nullptr || trajectory->trajectory_point().empty()) {
+    AERROR << "Planning trajectory not ready.";
+    return;
+  }
+
+  auto& logger = manager->LogBuffer();
+  const double distance = DistanceToPNCJunction(*trajectory);
+  const bool close_to_junction = distance >= 0;
```

Comment:
The logic here is in junction, not close to junction right ?

Reply:
`distance == 0` could be taken as "in junction".